BRKey.c

```diff
@@ -284,6 +284,33 @@ size_t BRKeyAddress(BRKey *key, char *addr, size_t addrLen) {
     return addrLen;
 }
 
+// writes the pay-to-witness-pubkey-hash nested in pay-to-script-hash (P2SH-P2WPKH) bitcoin address for key to addr:
+// https://bitcoincore.org/en/segwit_wallet_dev/#creation-of-p2sh-p2wpkh-address
+// returns the number of bytes written, or addrLen needed if addr is NULL
+size_t BRKeyWitnessAddress(BRKey *key, char *addr, size_t addrLen) {
+    // Only create witness addresses of compressed public keys
+    assert(key->compressed);
+
+    UInt160 pubKeyHash = BRKeyHash160(key);
+    uint8_t data[21], redeemScript[1 + BRScriptPushData(NULL, 0, pubKeyHash.u8, sizeof(pubKeyHash))];
```

Comment:
perhaps this was intended, but the code looks weird.
Maybe this was meant to be?
`uint8_t data[21] = redeemScript[1 + BRScriptPushData(NULL, 0, pubKeyHash.u8, sizeof(pubKeyHash))];`

(can't tell as I don't have an IDE/build scripts to even know if my assignment makes sense)

Reply:
@gubatron **almost** every function in Breadwallet Core returns the size needed in memory to store it then the function immediately gets called again to do the actual work.

```
// writes a data push script element to script
// returns the number of bytes written, or scriptLen needed if script is NULL
size_t BRScriptPushData(uint8_t *script, size_t scriptLen, const uint8_t *data, size_t dataLen);
```
The code looks weird all over the place cause this pattern keeps repeating in confusing ways.

Reply:
Usually, it's only done only when declaring variables. so it should be `functionThatReturnsASizeOfAnArray`.

This is one of the most confusing ones of this pattern since it's accounting to the assignment of `OP_1` so the `1 + `.

But a more common use is this one:

```c
int BRBase58Tests() {
    int r = 1;
    char *s;

    s = "#&$@*^(*#!^"; // test bad input

    uint8_t buf1[BRBase58Decode(NULL, 0, s)];
    size_t len1 = BRBase58Decode(buf1, sizeof(buf1), s);

    if (len1 != 0) r = 0, test_log("***FAILED*** %s: BRBase58Decode() test 1\n", __func__);
```

From BRTest.c

Reply:
oh I see, the array is being declared right there, duh, therefore defining its size.
For some reason I thought this was an existing variable. Brainfart.

Reply:
your rewrite of `BRTransactionVSize` seems to ignore a bunch of checks they do when adding up the input sizes, maybe that's screwing you up, or your tests show the estimated sizes end up being the same as theirs?

for instance. does your simplified for loop consider what happens when you get `OP_HASH160`, breadwallet's version adds a hard-coded size of prefix count of 23 bytes to those inputs

```
else if (input->script && input->scriptLen > 0 && input->script[0] == OP_HASH160) { // est. P2SH-P2WPKH sig size
            size += 23;
            witSize += TX_INPUT_SIZE;
}
```

Would in that case either `input->scriptLen` or `input->sigLen` or their addition be equal to 23. It seems like your function gets radically different sizes than theirs.

Reply:
Yeah theirs was very off, my function is tested on BRTest.c and passes with my implementation.

Unfortunately we only need the size to calculate fees this won’t help fixing the signing but it’s correct regardless.

Honestly I have no clue why he saw the VSize implementation as similar as the old Size was since they’re not!

The old function for size isn’t used anymore on this branch.